pkg/oci/httpreadseeker.go

```
@@ -0,0 +1,124 @@
+package oci
+
+import (
+	"io"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/errdefs"
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+)
+
+const maxRetry = 3
+
+type httpReadSeeker struct {
+	rc                 content.ReaderAt
+	log                logr.Logger
+	size               int64
+	offset             int64
+	errsWithNoProgress int
+	closed             bool
+}
+
+var _ io.ReadSeekCloser = (*httpReadSeeker)(nil)
+
+func newHTTPReadSeeker(log logr.Logger, reader content.ReaderAt) (io.ReadSeekCloser, error) {
+	if reader == nil {
+		return nil, errors.Errorf("httpReadSeeker: reader cannot be nil")
+	}
+	return &httpReadSeeker{
+		log:  log,
+		rc:   reader,
+		size: reader.Size(),
+	}, nil
+}
+
+func (hrs *httpReadSeeker) Read(p []byte) (n int, err error) {
+	if hrs.closed {
+		return 0, io.EOF
+	}
+
+	n, err = hrs.rc.ReadAt(p, hrs.offset)
+	hrs.offset += int64(n)
+	if n > 0 || err == nil {
+		hrs.errsWithNoProgress = 0
+	}
+	if err == io.ErrUnexpectedEOF {
+		// connection closed unexpectedly. try reconnecting.
+		if n == 0 {
+			hrs.errsWithNoProgress++
+			if hrs.errsWithNoProgress > maxRetry {
+				return // too many retries for this offset with no progress
+			}
+		}
+		if hrs.rc != nil {
+			if clsErr := hrs.rc.Close(); clsErr != nil {
+				hrs.log.Error(clsErr, "httpReadSeeker: failed to close ReadCloser")
```

We dont need the logger. We should just return the error.